PubSub Subscriber throws StatusRuntimeException when holding many leases

#### Environment details

1. OS type and version: max os x 10.16; linux 5.4.170+ (osImage: Container-Optimized OS from Google)
2. Java version: Oracle Java 11.0.10 on mac os; openjdk version "11.0.15" 2022-04-19 in docker/gke
3.  version(s): google-cloud-pubsub 1.116.0, 1.116.1, 1.116.2, 1.116.3, 1.116.4, 1.117.0

#### Steps to reproduce

  1. create pubsub topic and subscription
  2. publish at least 2775 messages to the topic
  3. use `com.google.cloud.pubsub.v1.Subscriber` to consume at least 2775 messages and hold their lease for at least 15 seconds.
  4. observe stack trace below

#### Code example

```java
import com.google.api.core.ApiFuture;
import com.google.api.gax.batching.FlowControlSettings;
import com.google.cloud.ServiceOptions;
import com.google.cloud.pubsub.v1.MessageReceiver;
import com.google.cloud.pubsub.v1.Publisher;
import com.google.cloud.pubsub.v1.Subscriber;
import com.google.cloud.pubsub.v1.SubscriptionAdminClient;
import com.google.cloud.pubsub.v1.TopicAdminClient;
import com.google.protobuf.ByteString;
import com.google.pubsub.v1.ProjectSubscriptionName;
import com.google.pubsub.v1.PubsubMessage;
import com.google.pubsub.v1.PushConfig;
import com.google.pubsub.v1.Subscription;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

public class HoldLeases {

  public static void main(String[] args) throws Exception {

    final String projectId = ServiceOptions.getDefaultProjectId();
    final TopicAdminClient topicAdminClient = TopicAdminClient.create();
    final SubscriptionAdminClient subscriptionAdminClient = SubscriptionAdminClient.create();
    final Subscription subscription = Subscription.newBuilder()
        .setName("projects/" + projectId + "/subscriptions/test-subscription-" + UUID.randomUUID())
        .setTopic("projects/" + projectId + "/topics/test-topic-" + UUID.randomUUID())
        .setAckDeadlineSeconds(10).build();

    topicAdminClient.createTopic(subscription.getTopic());
    subscriptionAdminClient.createSubscription(subscription.getName(), subscription.getTopic(),
        PushConfig.getDefaultInstance(), 0);
    final Publisher publisher = Publisher.newBuilder(subscription.getTopic()).build();

    try {
      // publish enough messages to exceed a single ModifyAckDeadline request
      int sentCount = 2775;
      List<ApiFuture<String>> published = new ArrayList<>(sentCount);
      for (int i = 0; i < sentCount; i++) {
        published.add(publisher
            .publish(PubsubMessage.newBuilder().setData(ByteString.copyFromUtf8("" + i)).build()));
      }
      publisher.publishAllOutstanding();
      for (ApiFuture<String> future : published) {
        future.get();
      }

      // pull all messages and hold them for 1 minute
      final CompletableFuture<Void> done = new CompletableFuture<Void>().orTimeout(15,
          TimeUnit.SECONDS);
      final MessageReceiver receiver = (message, consumer) -> {
        done.whenComplete((result, exception) -> {
          if (exception == null) {
            consumer.ack();
          } else {
            consumer.nack();
          }
        });
      };

      final Subscriber subscriber = Subscriber
          .newBuilder(ProjectSubscriptionName.parse(subscription.getName()), receiver)
          .setFlowControlSettings(
              FlowControlSettings.newBuilder().setMaxOutstandingElementCount(2775L)
                  .setMaxOutstandingRequestBytes(30_000_000L).build())
          .build();
      done.whenComplete((v, e) -> subscriber.stopAsync());
      try {
        subscriber.startAsync();
        subscriber.awaitTerminated();
      } finally {
        subscriber.stopAsync();
      }
    } finally {
      topicAdminClient.deleteTopic(subscription.getTopic());
      subscriptionAdminClient.deleteSubscription(subscription.getName());
      publisher.shutdown();
    }
  }
}
```

#### Stack trace
```
May 11, 2022 11:56:01 AM com.google.cloud.pubsub.v1.StreamingSubscriberConnection$2 onFailure
WARNING: failed to send operations
com.google.api.gax.rpc.InvalidArgumentException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Request payload size exceeds the limit: 524288 bytes.
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:92)
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:41)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:86)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:67)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1132)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1270)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1038)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:808)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:563)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:535)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:562)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:743)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:722)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Request payload size exceeds the limit: 524288 bytes.
	at io.grpc.Status.asRuntimeException(Status.java:535)
	... 17 more
```

#### Any additional information below

Issue does not appear with library version 1.115.5 and below.

I don't see the issue with less than 2772 messages in flight, but it's didn't start reliably showing until 2775 messages in flight.

--- a/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java
+++ b/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java
@@ -431,9 +431,9 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
     // Send modacks
     int pendingOperations = 0;
     for (ModackRequestData modackRequestData : modackRequestDataList) {
-      List<String> ackIdsInRequest = new ArrayList<>();
       for (List<AckRequestData> ackRequestDataInRequestList :
           Lists.partition(modackRequestData.getAckRequestData(), MAX_PER_REQUEST_CHANGES)) {
+        List<String> ackIdsInRequest = new ArrayList<>();
         for (AckRequestData ackRequestData : ackRequestDataInRequestList) {
           ackIdsInRequest.add(ackRequestData.getAckId());
           if (ackRequestData.hasMessageFuture()) {
@@ -511,9 +511,10 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
         // Remove from our pending operations
         ackOperationsWaiter.incrementPendingCount(-1);
 
+        Level level = isAlive() ? Level.WARNING : Level.FINER;
+        logger.log(level, "failed to send operations", t);
+
         if (!getExactlyOnceDeliveryEnabled()) {
-          Level level = isAlive() ? Level.WARNING : Level.FINER;
-          logger.log(level, "failed to send operations", t);
           return;
         }
 
@@ -578,9 +579,6 @@ final class StreamingSubscriberConnection extends AbstractApiService implements 
               currentBackoffMillis,
               TimeUnit.MILLISECONDS);
         }
-
-        Level level = isAlive() ? Level.WARNING : Level.FINER;
-        logger.log(level, "failed to send operations", t);
       }
     };
   }

--- a/google-cloud-pubsub/src/test/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnectionTest.java
+++ b/google-cloud-pubsub/src/test/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnectionTest.java
@@ -26,6 +26,7 @@ import com.google.api.gax.batching.FlowController;
 import com.google.api.gax.core.Distribution;
 import com.google.api.gax.rpc.StatusCode;
 import com.google.cloud.pubsub.v1.stub.SubscriberStub;
+import com.google.common.collect.Lists;
 import com.google.protobuf.Any;
 import com.google.pubsub.v1.AcknowledgeRequest;
 import com.google.pubsub.v1.ModifyAckDeadlineRequest;
@@ -71,8 +72,8 @@ public class StreamingSubscriberConnectionTest {
       "TRANSIENT_FAILURE_SERVICE_UNAVAILABLE";
   private static final String PERMANENT_FAILURE_OTHER = "I_DO_NOT_MATCH_ANY_KNOWN_ERRORS";
 
-  private static int MOCK_ACK_EXTENSION_DEFAULT = 10;
-  private static Duration ACK_EXPIRATION_PADDING_DEFAULT = Duration.ofSeconds(10);
+  private static int MOCK_ACK_EXTENSION_DEFAULT_SECONDS = 10;
+  private static Duration ACK_EXPIRATION_PADDING_DEFAULT_DURATION = Duration.ofSeconds(10);
   private static int MAX_DURATION_PER_ACK_EXTENSION_DEFAULT_SECONDS = 10;
 
   @Before
@@ -105,7 +106,8 @@ public class StreamingSubscriberConnectionTest {
 
     ModackRequestData modackRequestDataSuccess =
         new ModackRequestData(
-            MOCK_ACK_EXTENSION_DEFAULT, AckRequestData.newBuilder(MOCK_ACK_ID_SUCCESS).build());
+            MOCK_ACK_EXTENSION_DEFAULT_SECONDS,
+            AckRequestData.newBuilder(MOCK_ACK_ID_SUCCESS).build());
     modackRequestDataList.add(modackRequestDataSuccess);
 
     ModackRequestData modackRequestDataNack =
@@ -138,7 +140,8 @@ public class StreamingSubscriberConnectionTest {
     Map<String, String> errorInfoMetadataMapInitialRequest = new HashMap<String, String>();
     List<ModackRequestData> modackRequestDataList = new ArrayList<ModackRequestData>();
 
-    ModackRequestData modackRequestDataDefault = new ModackRequestData(MOCK_ACK_EXTENSION_DEFAULT);
+    ModackRequestData modackRequestDataDefault =
+        new ModackRequestData(MOCK_ACK_EXTENSION_DEFAULT_SECONDS);
 
     // Nack SUCCESS
     SettableApiFuture<AckResponse> messageFutureSuccessExpected = SettableApiFuture.create();
@@ -218,14 +221,14 @@ public class StreamingSubscriberConnectionTest {
         ModifyAckDeadlineRequest.newBuilder()
             .setSubscription(MOCK_SUBSCRIPTION_NAME)
             .addAllAckIds(ackIdsInitialRequest)
-            .setAckDeadlineSeconds(MOCK_ACK_EXTENSION_DEFAULT)
+            .setAckDeadlineSeconds(MOCK_ACK_EXTENSION_DEFAULT_SECONDS)
             .build();
 
     ModifyAckDeadlineRequest modifyAckDeadlineRequestRetry =
         ModifyAckDeadlineRequest.newBuilder()
             .setSubscription(MOCK_SUBSCRIPTION_NAME)
             .addAllAckIds(ackIdsRetryRequest)
-            .setAckDeadlineSeconds(MOCK_ACK_EXTENSION_DEFAULT)
+            .setAckDeadlineSeconds(MOCK_ACK_EXTENSION_DEFAULT_SECONDS)
             .build();
 
     // Set mock grpc responses
@@ -450,6 +453,53 @@ public class StreamingSubscriberConnectionTest {
         });
   }
 
+  @Test
+  public void testMaxPerRequestChanges() {
+    // Setup mocks
+    List<ModackRequestData> modackRequestDataList = new ArrayList<ModackRequestData>();
+    List<AckRequestData> ackRequestDataList = new ArrayList<AckRequestData>();
+
+    int numAckIds = 3000;
+    int numMaxPerRequestChanges = 1000;
+
+    List<String> mockAckIds = new ArrayList<String>();
+
+    for (int i = 0; i < numAckIds; i++) {
+      String mockAckId = "MOCK-ACK-ID-" + i;
+      mockAckIds.add(mockAckId);
+      ackRequestDataList.add(AckRequestData.newBuilder(mockAckId).build());
+    }
+
+    modackRequestDataList.add(
+        new ModackRequestData(MOCK_ACK_EXTENSION_DEFAULT_SECONDS, ackRequestDataList));
+
+    // Instantiate class and run operation(s)
+    StreamingSubscriberConnection streamingSubscriberConnection =
+        getStreamingSubscriberConnection(false);
+    streamingSubscriberConnection.sendAckOperations(ackRequestDataList);
+    streamingSubscriberConnection.sendModackOperations(modackRequestDataList);
+
+    // Assert expected behavior
+    for (List<String> mockAckIdsInRequest : Lists.partition(mockAckIds, numMaxPerRequestChanges)) {
+      AcknowledgeRequest expectedAcknowledgeRequest =
+          AcknowledgeRequest.newBuilder()
+              .setSubscription(MOCK_SUBSCRIPTION_NAME)
+              .addAllAckIds(mockAckIdsInRequest)
+              .build();
+      verify(mockSubscriberStub.acknowledgeCallable(), times(1))
+          .futureCall(expectedAcknowledgeRequest);
+
+      ModifyAckDeadlineRequest expectedModifyAckDeadlineRequest =
+          ModifyAckDeadlineRequest.newBuilder()
+              .setSubscription(MOCK_SUBSCRIPTION_NAME)
+              .addAllAckIds(mockAckIdsInRequest)
+              .setAckDeadlineSeconds(MOCK_ACK_EXTENSION_DEFAULT_SECONDS)
+              .build();
+      verify(mockSubscriberStub.modifyAckDeadlineCallable(), times(1))
+          .futureCall(expectedModifyAckDeadlineRequest);
+    }
+  }
+
   private StreamingSubscriberConnection getStreamingSubscriberConnection(
       boolean exactlyOnceDeliveryEnabled) {
     StreamingSubscriberConnection streamingSubscriberConnection =
@@ -466,7 +516,7 @@ public class StreamingSubscriberConnectionTest {
       StreamingSubscriberConnection.Builder builder) {
     return builder
         .setSubscription(MOCK_SUBSCRIPTION_NAME)
-        .setAckExpirationPadding(ACK_EXPIRATION_PADDING_DEFAULT)
+        .setAckExpirationPadding(ACK_EXPIRATION_PADDING_DEFAULT_DURATION)
         .setAckLatencyDistribution(mock(Distribution.class))
         .setSubscriberStub(mockSubscriberStub)
         .setChannelAffinity(0)

